### PR TITLE
Techdocs: Require access to related catalog entities (v2)

### PR DESCRIPTION
**Pull request by joeporpeglia**
## Hey, I just made a Pull Request!

Follow up from #8744.

Authorization of access to techdocs will be based on existing catalog permissions (i.e. if a user has read access for a given component, they will have access to the component's techdocs). Requesting the related entity from the catalog will implicitly check the given user's permissions. Doing this implicitly removes the need to make techdocs aware of the permission system and allows policy authors to focus primarily on catalog-based permissions. In the future, it may be beneficial to provide more granular control of access to techdocs.

- All endpoints, including `/static/docs` if `permission.enabled = true` in the config, will now request the related entity from the catalog to verify that the entity exists and the current user can access it. This was already the case for some endpoints that used entity data.
- Adds a `CachedEntityLoader` to optimize scenarios where loading a techdocs page results in many backend calls. For example, a page with 10 images will call the `/static/docs` endpoint for each one, resulting in unnecessary load on the catalog backend. This cache should have a very short TTL, enough to handle the requests within a single page load.
- `options.cache` is now required in router options, but the techdocs content cache will only be used when the `techdocs.cache.ttl` config value is present.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (iamEAP) [inline plugins/techdocs-backend/src/service/CachedEntityLoader.ts:81]: Q: does this also make network calls to the auth backend, or is this just pulling the identifier from the token in code?

File: plugins/techdocs-backend/src/service/router.ts
Comment (iamEAP) [inline plugins/techdocs-backend/src/service/router.ts:246]: Seems like the order of the middlewares here is quite important.  Might be worth a few test cases to prevent auth regressions on the static objects (e.g. object is cached but user is not auth'd).

File: plugins/techdocs-backend/src/service/router.ts
Comment (iamEAP) [inline plugins/techdocs-backend/src/service/router.ts:260]: I'm still thinking of the edge(-ish) case where folks have custom object storage schemas in place (supported by custom TechDocsStorage API implementations in the front-end).  Is there an existing mechanism to short-circuit this in such a case?  (e.g. via some `app-config.yaml` value, etc)

**Review by iamEAP — COMMENTED** [review]
Nice!  Couple of questions and suggestions, but the approach seems good, overall.  And also good with requiring the cache manager in the `createRouter` call (I believe it's been in `create-app` for quite some time, now).

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:57]: Just that the router wouldn't do anything differently / doesn't need that information. I guess it's similar to the catalog client handling 404s internally. Happy to move this logic if you have other thoughts though

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:48]: Good call! Does this seem like something to push down to the memcache client eventually?

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:81]: It pulls the data out of the token, but in the worst case it might result in a call to the auth backend to refresh the key store. See https://github.com/backstage/backstage/blob/a426ba0e26f3596126913c6793cf53eaaa3b25e3/plugins/auth-backend/src/identity/IdentityClient.ts#L131

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (iamEAP) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:48]: I think on @freben's advice, we shipped something simple to begin so we could see how usage evolved. It makes sense closer to the core, yeah.

File: plugins/techdocs-backend/src/service/router.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.ts:260]: There's nothing in place at the moment. I was originally thinking that users could connect `Publisher.docsRouter()` to a different path, like `/static/custom-docs`, but then they can't take advantage of the techdocs cache. It's also easy enough to toggle the behavior with a config value if that's preferred though!

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (iamEAP) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:58]: I'm imagining sometime in the future where someone else, without the context of this change, uses the `CachedEntityLoader` in this plugin, but hits a corner case where the expectation is that this is just a more performant way to load catalog entities, but the extra logic in here related to authz trips them up somehow...?  Maybe a stretch.

Could just be a matter of renaming this to make the purpose/intention a little clearer (like `CachedEntityLoaderForAuth` or something).  Or making this a more transparent cache and handling the auth-specific logic elsewhere.  Either way.

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (iamEAP) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:91]: Got it.  And if we were instead to use some derivation of the raw token value itself, is the worst-case just that sometime in the 5s TTL of the cached value, the token changes on the client-side, a new token gets passed instead, and there's a cache miss?  Or are there other implications?

File: plugins/techdocs-backend/src/service/router.ts
Comment (iamEAP) [reply plugins/techdocs-backend/src/service/router.ts:260]: Yeah, I'm especially curious on this because we as users of the plugin happen to fall into this category 😬 .

I was hoping there'd be some existing global-ish config that indicates "authz is explicitly enabled" and we can reliably, conditionally add the middleware in its presence?  Or is a TechDocs-specific exclude the better option?

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (iamEAP) [inline plugins/techdocs-backend/src/service/CachedEntityLoader.ts:93]: Fresh pot of coffee and I'm thinking about this in more detail: the cached objects here are sharing key-space with the existing object cache...  What's the risk of cache key collision?

Maybe low because of the `:` in there.  But could be worth an explicit `catalog:` prefix too?

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:58]: Hmm thinking on this more, perhaps its better to let the 403 errors bubble up and use the `errorHandler` middleware in the techdocs backend? That would enable more specific error messages on the frontend too

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:91]: Yup! Unless there's some security concern with using tokens in the cache key..

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:93]: Yeah good call, I'll add that prefix

File: plugins/techdocs-backend/src/service/router.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.ts:260]: There is a `permission.enabled` config value, but that applies to the authorization framework as a whole. I wonder if this should somehow be encoded in the `Publisher` interface. Serving "non-catalog" techdocs requires a custom publisher implementation anyway, so having a way to determine this might be better / less redundant than a new config field. Something like `techdocs.authorizeStaticAssets = false` is certainly lower friction to implement though 😄

File: plugins/techdocs-backend/src/service/router.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.ts:246]: Pushed a couple tests in 17b9542094dc0a10ff04f2cfc62a8d7475a5ad58

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (iamEAP) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:58]: Not sure where y'all landed on the 403/404 in order to not leak the existence of things, but I'd be down with that.

File: plugins/techdocs-backend/src/service/router.ts
Comment (iamEAP) [reply plugins/techdocs-backend/src/service/router.ts:260]: > Edit: Perhaps we can look at the config value for techdocs.publisher.type and enable the behavior when it's one of the built-in options?

I'm not sure the type is going to be a reliable indicator; it controls the canonical data store and configs but not necessarily anything else. For instance, we use the vanilla GCS publisher for serving static assets, we just happen to have an entirely separate process for actually _publishing_ to the GCS bucket (external to backstage completely) at the moment.

Maybe a boolean flag like you mentioned is the only option? But maybe it'd be useful to share one config for all plugins? Not sure.

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts]: Yeah both seem unlikely and should still respond successfully, just a bit slower. We might as well simplify things and remove the identity client dependency

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:58]: Ah just looking at #8711 now and this endpoint should return a 404 if the user doesn't have access. Removing this error handling logic!

File: plugins/techdocs-backend/src/service/router.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.ts:254]: What about a `publicDocsPath` config which this middleware can check to skip the catalog call? So if `publicDocsPath` matches the request path after `/static/docs`, then call `next()` and return early. The downside is that users will probably need to move things around in storage, but I like that it can work without disable techdocs authorization entirely

Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.ts:254]: I'll also add a check here so we only add the middleware if `permission.enabled = true`, which seems like a good optimization regardless of the custom docs setup

Comment (iamEAP) [reply plugins/techdocs-backend/src/service/router.ts:254]: > What about a publicDocsPath config which this middleware can check to skip the catalog call?

This is super interesting! Could even be useful in a default setup (eg make all docs under a namespace public?) but maybe that's complicated with the other metadata calls. Probably best to hold off until use-cases become clearer.

I'd be good with just the conditional middleware include on `permissions.enabled`.

Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.ts]: > Could even be useful in a default setup (eg make all docs under a namespace public?)

Smart! I like that idea, but you're right we should probably think more these use cases before introducing some "public docs" concept. I'll make a note to revisit this after.

> I'd be good with just the conditional middleware include on permissions.enabled.

Done in the latest commit

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [inline plugins/techdocs-backend/src/service/CachedEntityLoader.ts:75]: The token should probably be hashed or something before being used as a cache key. Any opinions on a specific approach here?

Comment (iamEAP) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:75]: Hmmm... I remember adding some key hashing logic somewhere in the core backend (to account for cache stores like memcache with limited key sizes). Don't remember if it was _only_ in the memcache case or for everything. But if it's everything, probably no point using cycles on additional hashing (assuming the token is quite long).

**Review by iamEAP — COMMENTED** [review]
Couple of last tiny little comments, but otherwise LGTM from my POV!  Beyond those, just needs a `minor` changeset.

File: plugins/techdocs-backend/src/service/CachedEntityLoader.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/CachedEntityLoader.ts:75]: That seems good for now!

File: plugins/techdocs-backend/src/service/router.ts
Comment (nissayeva) [inline plugins/techdocs-backend/src/service/router.ts:102]: would the cache not get set if there is no `defaultTtl`?

Comment (nissayeva) [reply plugins/techdocs-backend/src/service/router.ts:102]: oh i see but the entity cache and the techdocs cache are kept separate, never mind!

File: .changeset/giant-bees-speak.md
Comment (nissayeva) [inline .changeset/giant-bees-speak.md:7]: it might be helpful to connect this change back to why the cache is now required, it seems like a bit random otherwise if only looking at the changeset.

maybe even just more or less copy pasting the comment you have below:
```
// Entities are cached to optimize the /static/docs request path, which can be called many times
// when loading a single techdocs page.
```

**Review by nissayeva — APPROVED** [review]
my approval means nothing but lgtm!

File: .changeset/giant-bees-speak.md
Comment (iamEAP) [inline .changeset/giant-bees-speak.md:9]: Maybe a more generic phrasing of this; docs without linked catalog entities may be one use-case, but there are likely others.  Perhaps:

> Note: If you publish your TechDocs documentation under custom paths that do not conform to the default `:namespace/:kind/:name` pattern, TechDocs will not work with permissions enabled....

Comment (joeporpeglia) [reply .changeset/giant-bees-speak.md:9]: Good point! I also added a note about joining the discord to help us better understand these custom use cases. I saw some other changesets link to discord so I figured that was okay 😄

**Comment by joeporpeglia** [comment]
@iamEAP we plan on doing a bigger documentation push after we've finished the core plugin integrations. It might be confusing to mention authorization in the techdocs documentation before we have that general framework documented. Behavior _should_ remain the same with permissions disabled, but lmk if there's anything you think we should call out before we get around to those docs!

**Review by iamEAP — APPROVED** [review]
Excellent!  Let's do this.  Thanks again, @joeporpeglia.

File: plugins/techdocs-backend/src/service/router.test.ts
Comment (freben) [inline plugins/techdocs-backend/src/service/router.test.ts:149]: This prototype usage confounds me 😄 Never seen that before.

Anyway, mocking the config reader sounds inadvisable. Just giving in actual config instances with `new MockConfigApi({ techdocs: { cache: { ttl: 1 }}})` etc feels more appropriate.

**Review by freben — APPROVED** [review]
Approving to unblock!

File: plugins/techdocs-backend/src/service/router.test.ts
Comment (joeporpeglia) [reply plugins/techdocs-backend/src/service/router.test.ts:149]: Agreed! This test suite was already using this prototype pattern so I just stuck with that. Happy to do this refactor with the follow up work we discussed in https://github.com/backstage/backstage/pull/8791#discussion_r780117326

**Pull request merged**
